h/static/scripts/websocket.js

```diff
@@ -32,44 +38,80 @@ function Socket(url) {
     }
   }
 
-  function reconnect() {
-    var didConnect = false;
-    var connectOperation = retry.operation();
-    connectOperation.attempt(function (currentAttempt) {
+  // Connect the websocket immediately. If a connection attempt is already in
+  // progress, do nothing.
+  function connect() {
+    if (operation) {
+      return;
+    }
+
+    operation = retry.operation({
+      minTimeout: RECONNECT_MIN_DELAY * 2,
+      // Don't retry forever -- fail permanently after 10 retries
+      retries: 10,
+      // Randomize retry times to minimise the thundering herd effect
+      randomize: true
+    });
+
+    operation.attempt(function () {
       socket = new WebSocket(url);
       socket.onopen = function (event) {
-        // signal successful connection
-        connectOperation.retry();
-        didConnect = true;
-        sendMessages();
-
+        onOpen();
         self.emit('open', event);
       };
-
       socket.onclose = function (event) {
-        if (event.code !== CLOSE_NORMAL) {
-          if (didConnect) {
-            console.warn('The WebSocket connection closed abnormally ' +
-              '(code: %d, reason: %s). Reconnecting automatically.',
-              event.code, event.reason);
-            reconnect();
-          } else {
-            console.warn('Retrying connection (attempt %d)', currentAttempt);
-            connectOperation.retry(new Error(event.reason));
-          }
+        if (event.code === CLOSE_NORMAL) {
+          self.emit('close', event);
+          return;
         }
-        self.emit('close', event);
+        var err = new Error('WebSocket closed abnormally, code: ' + event.code);
+        console.warn(err);
+        onAbnormalClose(err);
       };
-
       socket.onerror = function (event) {
         self.emit('error', event);
       };
-
       socket.onmessage = function (event) {
         self.emit('message', event);
       };
     });
-  };
+  }
+
+  // Reconnect the websocket after a short delay. This is used when the
+  // websocket has closed abnormally. If reconnection is already in progress, do
+  // nothing.
+  function reconnect() {
+    if (operation) {
```

Comment:
We're checking if an operation is in progress both here and in `onAbnormalClose`. Given that we only call `reconnect()` in one place, I think it would probably be clearer just to put this logic in `onAbnormalClose`.